runtime/parachains/src/disputes.rs

```
@@ -104,20 +105,66 @@ impl PunishValidators for () {
 	/// Whether the chain is frozen, if the chain is frozen it will not accept
 	/// any new parachain blocks for backing or inclusion.
 	fn is_frozen() -> bool;
 
-	/// Handler for filtering any dispute statements before including them as part
-	/// of inherent data. This can be useful to filter out ancient and duplicate
-	/// dispute statements.
-	fn filter_multi_dispute_data(statement_sets: &mut MultiDisputeStatementSet);
+	/// Remove dispute statement duplicates and sort the non-duplicates based on
+	/// local (lower indicies) vs remotes (higher indices) and age (older with lower indices).
+	fn deduplicate_and_sort_dispute_data(
+		statement_sets: &mut MultiDisputeStatementSet,
+	) -> Result<(), ()> {
+		use core::cmp::Ordering;
+
+		// TODO: Consider trade-of to avoid `O(n * log(n))` average lookups of `included_state`
+		// TODO: instead make a single pass and store the values.
+		let n = statement_sets.len();
+		// Sort the dispute statements according to the following prioritization:
+		//  1. Prioritize local disputes over remote disputes.
+		//  2. Prioritize older disputes over newer disputes.
+		statement_sets.sort_by(|a: &DisputeStatementSet, b: &DisputeStatementSet| {
+			let a_local_block = Self::included_state(a.session, a.candidate_hash);
+			let b_local_block = Self::included_state(b.session, b.candidate_hash);
+			match (a_local_block, b_local_block) {
+				// Prioritize local disputes over remote disputes.
+				(None, Some(_)) => Ordering::Greater,
+				(Some(_), None) => Ordering::Less,
+				// For local disputes, prioritize those that occur at an earlier height.
+				(Some(a_height), Some(b_height)) => a_height.cmp(&b_height),
+				// Prioritize earlier remote disputes using session as rough proxy.
+				(None, None) => a.session.cmp(&b.session),
+			}
+		});
+		statement_sets.dedup();
+
+		// if there were any duplicates, indicate that to the caller.
+		if n == statement_sets.len() {
+			Ok(())
+		} else {
+			Err(())
```

Would be worth explaining why duplicates are an error and how it is handled. IMO finding duplicates should not be an error condition, but I might be wrong.

---

The caller should be notified and is the one to decide how the handle it. The information should not be swallowed silently, it removes the option to act upon it.

---

Ok, I am guessing to make sure a block author is not trying to spam us with lots of duplicate disputes, where we would have to check signatures on each of those. I can see though, that we are ignoring the error on the calling side ... could we log it?

---

Yeah, that's a good point. Logging is the very least we should do. But this is not even a warning, it's only a `debug!`